resolwe/flow/managers/listener.py

```diff
@@ -0,0 +1,253 @@
+            'spawn_processors': [list of spawn dictionaries],
+            'exported_files_mapper': [dict with file mappings]
+        }
+        """
+        data_id = obj['id']
+        str_id = str(data_id)  # json only has string keys, so hydrate_spawned_files will need this stringified
+        exported_files_mapper = obj['exported_files_mapper']
+        with transaction.atomic():
+            parent_data = Data.objects.get(pk=data_id)
+
+            # Spawn processors
+            for d in obj['spawn_processors']:
+                d['contributor'] = parent_data.contributor
+                d['process'] = Process.objects.filter(slug=d['process']).latest()
+
+                for field_schema, fields in iterate_fields(d.get('input', {}), d['process'].input_schema):
+                    type_ = field_schema['type']
+                    name = field_schema['name']
+                    value = fields[name]
+
+                    if type_ == 'basic:file:':
+                        fields[name] = self.hydrate_spawned_files(exported_files_mapper, value, str_id)
+                    elif type_ == 'list:basic:file:':
+                        fields[name] = [self.hydrate_spawned_files(exported_files_mapper, fn, str_id) for fn in value]
+
+                with transaction.atomic():
+                    d = Data.objects.create(**d)
+                    d.parents.add(parent_data)
+
+                    # Copy permissions.
+                    copy_permissions(parent_data, d)
+
+                    # Copy collections.
+                    for collection in parent_data.collection_set.all():
+                        collection.data.add(d)
+
+        self._send_reply(obj, {'result': 'OK'})
+
+    def run(self):
+        """Start the main listening loop. Doesn't return until terminate() is called."""
+        # psycopg connection objects are thread-safe, but django makes them all thread-local,
+        # so we overwrite the new local connection with the one from the main thread,
+        # in order to participate in the same transactions (needed for testcases with transactions)
+        old_db = db.connections[db.DEFAULT_DB_ALIAS]
+        db.connections[db.DEFAULT_DB_ALIAS] = self._main_db
+        chan = ExecutorListener.channel_name_pair()[0]
+        while not self._terminated.is_set():
+            ret = self._redis.blpop(chan, timeout=1)
+            if ret is None:
+                continue
+            _, item = ret
+            obj = json.loads(item.decode('utf-8'))
+
+            if 'action' not in obj:
+                continue
+
+            handler = getattr(self, 'handle_' + obj['action'], None)
+            if handler:
+                try:
+                    handler(obj)
+                except:  # pylint: disable=bare-except
+                    logger.error(__("Executor command handling error:\n\n{}", traceback.format_exc()))
+            else:
+                pass  # what do we do in case of invalid messages?
```

Comment:
Maybe also include the message as an `extra` when logging, also above in case of errors. This can be helpful when inspecting errors, e.g. in Sentry.